anchore_engine/clients/grype_wrapper.py

```diff
@@ -447,79 +549,147 @@ def init_grype_db_engine(
+            self._staging_grype_db_session_maker = latest_grype_db_session_maker
 
-            # Remove the old local db only if it's not the current db
-            if old_grype_db_dir and old_grype_db_dir != self._grype_db_dir:
-                self._remove_local_grype_db(old_grype_db_dir)
+            # Return the staging engine metadata as a data object
+            return self.get_grype_db_engine_metadata(use_staging=True)
 
-    def _get_metadata_file_contents(self, metadata_file_name) -> json:
+    def unstage_grype_db(self):
         """
-        Return the json contents of one of the metadata files for the in-use version of grype db
+        Unstages the staged grype_db. This method returns the production grype_db engine metadata, if a production
+        grype_db has been set. Otherwise it returns None.
         """
-        # Get the path to the latest metadata file
-        latest_metadata_file = os.path.join(
-            self._grype_db_dir, self._grype_db_version, metadata_file_name
-        )
+        self._staging_grype_db_dir = None
+        self._staging_grype_db_version = None
+        self._staging_grype_db_session_maker = None
 
-        # Ensure the file exists
-        if not os.path.exists(latest_metadata_file):
-            # If not, return None
+        # Return the existing, production engine metadata as a data object
+        try:
+            return self.get_grype_db_engine_metadata(use_staging=False)
+        except ValueError as error:
+            logger.warn(
+                "Cannot return production grype_db engine metadata, as none has been set."
+            )
             return None
-        else:
-            # Get the contents of the file
-            with open(latest_metadata_file) as read_file:
-                try:
-                    return json.load(read_file)
-                except JSONDecodeError:
-                    logger.error(
-                        "Unable to decode metadata file into json: %s",
-                        read_file,
+
+    def update_grype_db(self, archive_checksum: str):
+        """
+        Checks to ensure a new grype_db has been staged, and raises a ValueError if it has not. Otherwise
+        this promotes the staged grype_db to the production grype_db, and unstages the previously-staged
+        grype_db.
+
+        To ensure that the caller is promoting the correct staged grype-db (ie the one it think it is
+        promoting to production, this method is parameterized with archive_checksum, which must be supplied and
+        match the archive_checksum stored in the staging engine metadata.
+        """
+        with self.write_lock_access():
+            # Ensure a grype_db has been staged, and raise an error if not.
+            if (
+                not self._staging_grype_db_dir
+                and not self._staging_grype_db_version
+                and not self._staging_grype_db_session_maker
+            ):
+                raise ValueError(
```

Comment:
i think it's better to make exceptions for all of these special cases instead of raising ValueError with different strings. this will allow the caller to better handle different scenarios.

Reply:
the exceptions can subclass ValueError